obfs4/common.go

```diff
@@ -7,36 +7,33 @@ import (
 	"net/url"
 )
 
-// Node is a proxy node, that can be used to construct a proxy chain.
-type Node struct {
-	Addr     string     // ag: I'm guessing this is used like ip:port
-	Host     string     // ... but then this is redundant
-	Protocol string     // obfs4 in this case
-	url      *url.URL   // url
+// ProxyNode is a proxy node, that can be used to construct a proxy chain.
+type ProxyNode struct {
+	Addr     string
+	Protocol string // obfs4 in this case
+	url      *url.URL
 	Values   url.Values // contains the cert and iat-mode parameters
-	//Transport string     // this only makes sense if/when we do use different transporters for obfs4. for the time being this can be removed, or perhaps denoted as "raw"
+	// base dialer to be passed to obfuscation dialer
+	Dial DialFunc
 }
 
-// NewNodeNewNodeFromURI returns a configured proxy node. It accepts a string with all the parameters
-// needed to establish a connection to the obfs4 proxy, in the form:
-// obfs4://<ip>:<port>?cert=<deadbeef>&iat-mode=<int>
-func NewNodeFromURI(uri string) (Node, error) {
+// NewProxyNodeFromURI returns a configured proxy node. It accepts a string
+// with all the parameters needed to establish a connection to the obfs4
+// proxy, in the form "obfs4://<ip>:<port>?cert=<deadbeef>&iat-mode=<int>"
+func NewProxyNodeFromURI(uri string) (ProxyNode, error) {
 	u, err := url.Parse(uri)
 	if err != nil {
-		return Node{}, err
+		return ProxyNode{}, err
 	}
 	log.Printf("Using %s proxy at %s:%s", u.Scheme, u.Hostname(), u.Port())
-	// q, err := url.ParseQuery(u.RawQuery)
-	// log.Println("cert:", url.QueryEscape(q["cert"][0]))
 
 	if u.Scheme != "obfs4" {
-		return Node{}, fmt.Errorf("expected obfs4:// uri")
+		return ProxyNode{}, fmt.Errorf("expected obfs4:// uri")
 	}
 
-	return Node{
+	return ProxyNode{
 		Protocol: u.Scheme,
 		Addr:     net.JoinHostPort(u.Hostname(), u.Port()),
```

Comment:
You do not seem to be checking whether a port is defined here, so what happens if you pass a URL that does not have any port? Also, `u.Host` is equivalent to `u.Hostname()` when there is a port.

Reply:
Good question, and good catch. It should fail if no port is specified, there's no way to assign a default port.

Reply:
added tests, and checked for missing port, hostname